csrc/multidevice/executor.cpp

```diff
@@ -6,133 +6,234 @@
-    }
-    // Run the stage to get concrete outputs or placeholders
-    // TODO: deal with aliases I/O. For example if the stage is empty, i.e.,
-    // Inputs=Outputs, we need to handle them anyway
-    outputs = fe_[stage]->runFusion(stage_input_IValues);
+  // Compile the group and execute it with FusionExecutor
+  // Check if the executor has been cached. If not, create and cache it
+  if (fe_.find(group) == fe_.end()) {
+    fe_.emplace(group, std::make_unique<FusionExecutor>());
+    fusions_.emplace(group, staged_fusion_->makeFusion(group));
+    fe_[group]->compileFusion(fusions_.at(group).get(), group_input_IValues);
   }
+  outputs = fe_[group]->runFusion(group_input_IValues);
 
-  // Store the outputs or placeholders in the context
+  // Store the outputs in the context
   for (auto output_idx : c10::irange(outputs.size())) {
-    val_to_IValue_
-        [stage->outputs().at(output_idx)->as<PipelineVal>()->getOriginalVal()] =
-            outputs.at(output_idx);
+    val_to_IValue_[group->outputs().at(output_idx)] = outputs.at(output_idx);
   }
 }
 
-void PipelineExecutor::handle(PipelineCommunication* c) {
-  auto input_val = c->in()->as<PipelineVal>()->getOriginalVal();
-  auto output_val = c->out()->as<PipelineVal>()->getOriginalVal();
-
-  at::Tensor input_tensor, output_tensor;
-  if (val_to_IValue_.find(input_val) != val_to_IValue_.end()) {
-    input_tensor = val_to_IValue_.at(input_val).toTensor();
-  }
-  if (val_to_IValue_.find(output_val) != val_to_IValue_.end()) {
-    output_tensor = val_to_IValue_.at(output_val).toTensor();
-  }
-
-  // Lower the Communication into a vector of Communications
-  if (communications_.find(c) == communications_.end()) { // check if cached
+void MultiDeviceExecutor::postCommunication(SegmentedGroup* group) {
+  // Lower the group into a vector of Communications
+  if (communications_.find(group) == communications_.end()) { // check if cached
+    NVF_ERROR(
+        group->exprs().size() == 1,
+        "Communication segments must contain only one Expr");
+    auto expr = group->exprs().at(0);
+    NVF_ERROR(
+        expr->inputs().size() == 1,
+        "Communication must have exactly one input");
+    NVF_ERROR(
+        expr->outputs().size() == 1,
+        "Communication must have exactly one output");
+    auto input_val = expr->inputs().at(0);
+    auto output_val = expr->outputs().at(0);
+    at::Tensor input_tensor, output_tensor;
+    if (val_to_IValue_.find(input_val) != val_to_IValue_.end()) {
+      input_tensor = val_to_IValue_.at(input_val).toTensor();
+    }
+    if (val_to_IValue_.find(output_val) != val_to_IValue_.end()) {
+      output_tensor = val_to_IValue_.at(output_val).toTensor();
+    }
     communications_.emplace(
-        c,
+        group,
         lowerCommunication(
-            runtime_.comm_.deviceId(), c, input_tensor, output_tensor));
+            comm_.deviceId(), expr, input_tensor, output_tensor));
```

Comment:
Does this cache input_tensor and output_tensor?

Reply:
I suppose Meghan is asking if the aten tensors are reused. I think they should not be reused, but looks like they are indeed reused, right?

Reply:
> I suppose Meghan is asking if the aten tensors are reused. I think they should not be reused, but looks like they are indeed reused, right?

Yes, they are reused. I agree with you that this is not ideal (and there might be a bug if we execute the same fusion multiple times, I just thought about it). However, this is the only way to use the cache for now because, at a lower level, the communications (in `multidevice/communications.h`) are instantiated with Aten buffers, instead of symbolic values.

If this is a concern we can simply turn off this basic caching mechanism for now.

Reply:
I actually removed the communication lowering's caching, I think it is faulty, and for sure it should be made better to be interesting. Will work on it on a separate PR